### PR TITLE
[WIP] State Tracing

### PR DESCRIPTION
**Pull request by insipx**
Preliminary state tracing support for native runtimes

related:
https://github.com/paritytech/substrate/pull/6916
https://github.com/paritytech/substrate/pull/5826
https://github.com/paritytech/substrate/pull/6672

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -44,30 +49,27 @@ codec = { package = "parity-scale-codec", version = "1.3", default-features = fa
...
-sc-chain-spec = { git = "https://github.com/paritytech/substrate", branch = "master", package = "sc-chain-spec" }
-sp-trie = { git = "https://github.com/paritytech/substrate", branch = "master", package = "sp-trie" }
-sp-state-machine = { git = "https://github.com/paritytech/substrate", branch = "master", package = "sp-state-machine" }
-sp-io = { git = "https://github.com/paritytech/substrate", branch = "master", package = "sp-io" }
-itoa = "0.4.6"
-include_dir = "0.6.0"
-tempfile = "3.1.0"
-async-executor = "0.1.2"
-arc-swap = "0.4.7"
+sp-database = { git = "https://github.com/paritytech/substrate", package="sp-database" , rev = "35fe3cd1bc4b64cadb0bc6196ae40173db65bb28" }
```
Comment (dvdplm) [inline Cargo.toml:52]: rev `35fe3cd1bc4b64cadb0bc6196ae40173db65bb28` is rather old (Aug 26) – what is stopping us from using `master` or `2.0`, i.e. what motivated the change?

File: bin/polkadot-archive/src/archive.rs [hunk 2]
```diff
@@ -23,6 +23,28 @@ use polkadot_service::westend_runtime as westend_rt;
 use polkadot_service::Block;
 use sc_chain_spec::ChainSpec;
 use substrate_archive::{Archive, ArchiveBuilder};
+use sc_executor::native_executor_instance;
+
+native_executor_instance!(
+	pub KsmExec,
+	ksm_rt::api::dispatch,
+    ksm_rt::native_version,
```
Comment (dvdplm) [inline bin/polkadot-archive/src/archive.rs:31]: funny whitespace

File: src/actors.rs [hunk 3]
```diff
@@ -222,8 +218,16 @@ where
...
+                    match t {
+                        Err(coil::FetchError::Timeout) => {
+                            log::warn!("Timeout waiting for task to start");
+                            continue;
+                        },
+                        v => {
+                            if v? == 0 {
+                                smol::Timer::new(std::time::Duration::from_millis(3600)).await;
+                            }
+                        }
```
Comment (dvdplm) [inline src/actors.rs:230]: Maybe we can do something like

```
Ok(v) if v == 0 => { … … },
Ok(_) => {} // but maybe we want to log something here?
Err(FailedLoadingJob) => { … … }
```

Not sure if it's better, but sometimes explicit is good.

File: src/actors/workers/state_tracing.rs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with substrate-archive.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::error::Result;
+use sc_tracing::{SpanDatum, TraceEvent, TraceHandler, ProfilingSubscriber};
+use xtra::prelude::*;
+
+const TRACE_TARGETS: &str = "assets,atomic-swap,aura,authority-discovery,authorship, \
```
Comment (dvdplm) [inline src/actors/workers/state_tracing.rs:21]: How do we obtain this list? Can we do anything smart here? Can we stick them in a toml file at least? Ideally we should have an RPC to ask the node about the supported targets.

File: src/actors/workers/state_tracing.rs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+impl TraceHandler for ArchiveTraceHandler {
+    fn handle_span(&self, sd: SpanDatum) {
+       log::info!("Got message");
+        println!("Outside");
+        if let Some(a) = self.addr.as_ref() {
+            println!("Inside");
+            let _ = a.do_send(SpanMessage(sd));
+        }
```
Comment (dvdplm) [inline src/actors/workers/state_tracing.rs:48]: funny whitespace and spurious `println`?

File: src/actors/workers/state_tracing.rs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    }
+}
+
+#[async_trait::async_trait]
+impl Actor for ArchiveTraceHandler {
+    async fn started(&mut self, ctx: &mut Context<Self>) {
+        let sub = ProfilingSubscriber::new_with_handler(Box::new(self.clone()), TRACE_TARGETS);
+        let addr = ctx.address().expect("Actor just started");
+        self.addr = Some(addr);
+        tracing::subscriber::set_global_default(sub).unwrap();
```
Comment (dvdplm) [inline src/actors/workers/state_tracing.rs:67]: Use `expect`

File: src/backend/frontend.rs [hunk 7]
```diff
@@ -125,8 +132,8 @@ fn execution_strategies() -> ExecutionStrategies {
     ExecutionStrategies {
         syncing: ExecutionStrategy::NativeElseWasm,
         importing: ExecutionStrategy::NativeElseWasm,
-        block_construction: ExecutionStrategy::NativeElseWasm,
-        offchain_worker: ExecutionStrategy::NativeWhenPossible,
+        block_construction: ExecutionStrategy::NativeWhenPossible,
```
Comment (dvdplm) [inline src/backend/frontend.rs:135]: dumb q: how is `NativeWhenPossible` different from `NativeElseWasm`?

**Review by dvdplm — CHANGES_REQUESTED** [review]
Left some nitpicks. I have a hard time seeing where this is heading and I could use a better description; it seems like this PR doesn't doesn't store the traces anywhere?
I am also a little concerned about having to fork polkadot for this – what do we need from upstream to reach a stable point? After all our end goal here is to provide a production quality state tracing tool so it's critical that we have a good maintenance story and a stable situation upstream. Can you elaborate a bit on what the plan is?

**Comment by insipx** [comment]
Closing in favor of #134

**Pull request closed without merging**